// image-processing/dist/test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var path = require("path");
var index_1 = require("./index");
var atlases = [
    {
        files: [
            {
                path: ['example-sprites/game/**/*.png'],
                convertOption: {
                    scaleX: 1,
                    scaleY: 1,
                    dontExtent: true,
                    maxWidth: 100,
                    maxHeight: 100,
                    trim: true
                }
            }
        ],
        layoutConfig: {
            max_width: 512,
            max_height: 512,
            padding: 10,
            oversized_warning: true,
        },
        exportConfig: {
            ext: 'png',
            quality: 1,
            compressor: null
        }
    }
];
var generateOptions = {
    concurrencyLimit: 10,
    cleanCache: false,
    cleanOutput: false,
    inputPath: path.resolve(__dirname, '../test/input'),
    outputPath: path.resolve(__dirname, '../test/output'),
    cachePath: path.resolve(__dirname, '../test/cache'),
    outputDestination: function (atlasIndex, pageIndex, res) {
        var sheet = res.atlases[atlasIndex].sheets[pageIndex];
        return sheet.hash + path.extname(sheet.path);
    },
    sriteNameResolver: function (atlasIndex, filePath, res) {
        var atlas = res.atlases[atlasIndex];
        return path.relative(generateOptions.inputPath, filePath);
    },
    atlasPageImagePathResolver: function (atlasIndex, pageIndex, res) {
        var sheet = res.atlases[atlasIndex].sheets[pageIndex];
        return path.relative(generateOptions.outputPath, sheet.path);
    },
    atlasPageMetaPathResolver: function (atlasIndex, pageIndex, res) {
        var sheet = res.atlases[atlasIndex].sheets[pageIndex];
        var name = path.basename(sheet.path);
        return generateOptions.outputPath + '/' + name + '.json';
    }
};
index_1.metapacker(atlases, generateOptions, console, {
    saver: 'PIXIManifestSaver',
    imageProcessor: 'ImageMagickImageProcessor'
});
